### PR TITLE
Utilities for splitting and unsplitting mode objects

### LINKED ISSUE #1938: Utilities to translate from `ReverseMode` to `ReverseModeSplit`
In DI, the `pullback` functions use split reverse mode by default, so they disregard the `mode` field of `AutoEnzyme` (because I figure no one is going to pass a `ReverseSplitWithPrimal` anyway). But still I'd like to respect some of the attributes of the `ReverseMode` that the user provides.
Would you welcome a PR with functions that translate a `ReverseMode` to the corresponding `ReverseModeSplit`, setting reasonable defaults for the remaining type parameters?

### PR DESCRIPTION
**Pull request by gdalle**
Fix #1938 

I can also add more tests for these mode modifiers in a separate PR, if you're interested

### CODE REVIEW

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (gdalle) [inline lib/EnzymeCore/src/EnzymeCore.jl:387]: I think the type parameters were switched. It has no concrete effect because the struct might as well be defined as
```julia
struct ReverseModeSplit{T1,T2,T3,...}
```
but fixing it makes reading the code easier

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (gdalle) [inline lib/EnzymeCore/src/EnzymeCore.jl:606]: I took the default values from `ReverseSplitWithPrimal`

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (wsmoses) [inline lib/EnzymeCore/src/EnzymeCore.jl:387]: Probably keep this the same?

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (wsmoses) [inline lib/EnzymeCore/src/EnzymeCore.jl:604]: Can these be (perhaps default) args to this fn?

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (wsmoses) [inline lib/EnzymeCore/src/EnzymeCore.jl:632]: Combined mode is usually what we refer to it as

File: lib/EnzymeCore/test/mode_modification.jl
Comment (wsmoses) [inline lib/EnzymeCore/test/mode_modification.jl:12]: Can you also test the use of runtime activity and err if find written

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (gdalle) [reply lib/EnzymeCore/src/EnzymeCore.jl:387]: As explained in [this comment](https://github.com/EnzymeAD/Enzyme.jl/pull/1979#discussion_r1805084780), this just fixes a typo that doesn't change functionality but makes the code harder to understand. Semantically, the runtime activity parameter comes before width in all the uses of this struct, even though they are "labelled" wrong here.

File: lib/EnzymeCore/test/mode_modification.jl
Comment (gdalle) [reply lib/EnzymeCore/test/mode_modification.jl]: done

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (gdalle) [reply lib/EnzymeCore/src/EnzymeCore.jl]: renamed it

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (gdalle) [reply lib/EnzymeCore/src/EnzymeCore.jl]: I added them as positional arguments as `Val` types with default values, to ensure type stability

File: lib/EnzymeCore/test/mode_modification.jl
Comment (wsmoses) [inline lib/EnzymeCore/test/mode_modification.jl:22]: Oh sorry one more, if you could also confirm that ABI is preserved too

**Review by wsmoses — APPROVED** [review]
One more test then LGTM!

File: lib/EnzymeCore/test/mode_modification.jl
Comment (gdalle) [reply lib/EnzymeCore/test/mode_modification.jl:22]: I tried that but `set_abi` isn't defined yet for `ReverseModeSplit`, want me to add it?

Comment (wsmoses) [reply lib/EnzymeCore/test/mode_modification.jl:22]: Yeah go for it, probably an oversight

Comment (gdalle) [reply lib/EnzymeCore/test/mode_modification.jl:23]: done

**Comment by gdalle** [comment]
Silly question: do the tests of EnzymeCore ever run? Because that's where I put these new checks

**Comment by wsmoses** [comment]
....good question. Probably we should just add a CI job for EnzymeCore

**Comment by gdalle** [comment]
See https://github.com/EnzymeAD/Enzyme.jl/pull/2009

**Comment by gdalle** [comment]
Since #2009 has been merged by @vchuravy, I have integrated the changes here and I think this one is good to go.

**Comment by gdalle** [comment]
Is anything more needed on my end?

**Comment by gdalle** [comment]
Gentle bump on this one @wsmoses

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (wsmoses) [inline lib/EnzymeCore/src/EnzymeCore.jl:424]: just noticed this is also missing a version for set_abi on the type [not just value], otherwise lgtm

**Review by wsmoses — APPROVED** [review]

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (gdalle) [reply lib/EnzymeCore/src/EnzymeCore.jl:424]: I'm not sure what you mean, the signature is
```
set_abi(::Mode, ::Type{A}) where {A<:ABI}
```
and it is implemented consistently for all modes. Do you mean you also want
```
set_abi(::Type{M}, ::Type{A}) where {M<:Mode,A<:ABI}
```
everywhere?

Comment (wsmoses) [reply lib/EnzymeCore/src/EnzymeCore.jl:424]: like this https://github.com/EnzymeAD/Enzyme.jl/commit/03d70353863e245ceae1dc04a201575073509be2

Comment (wsmoses) [reply lib/EnzymeCore/src/EnzymeCore.jl:424]: you add set_abi methods for split, and it should come with both type and mode

Comment (gdalle) [reply lib/EnzymeCore/src/EnzymeCore.jl:436]: done

File: lib/EnzymeCore/src/EnzymeCore.jl
Comment (wsmoses) [inline lib/EnzymeCore/src/EnzymeCore.jl:660]: @vchuravy should we have this be lowercase or upper case. I'm actually now thinking maybe uppercase. I like the WithPrimal(Forward) == ForwardWithPrimal property.

Comment (wsmoses) [reply lib/EnzymeCore/src/EnzymeCore.jl:660]: so then similarly here Split(ReverseWithPrimal) == ReverseSplitWithPrimal

Comment (gdalle) [reply lib/EnzymeCore/src/EnzymeCore.jl:660]: Whenever possible, it would be nice to keep [consistency with Base Julia](https://docs.julialang.org/en/v1/manual/style-guide/#Use-naming-conventions-consistent-with-Julia-base/): uppercase for types, lowercase for functions.
Of course this is already broken by some of the existing mode utilities, and it's too late to go back, but maybe adopting a minimal style guide would help? That way, contributors know right away how to comply, instead of following changes of hearts from maintainers.

**Comment by gdalle** [comment]
Hi @wsmoses @vchuravy, any updates on this? It got two approvals already but then more changes were requested every time.
Sorry to be insistent, it is just important for DI. Some operators use split mode instead of combined mode (e.g. pullback when the output is not a number). At the moment, for lack of translating utilities, DI completely [disregards the mode information](https://github.com/JuliaDiff/DifferentiationInterface.jl/blob/675b4bd8885b5b120f8640e47e90c370fa93f843/DifferentiationInterface/ext/DifferentiationInterfaceEnzymeExt/utils.jl#L63-L66) provided by the user in the backend. So even when they give a mode like `set_runtime_activity(Reverse)`, DI will fall back on the default `ReverseSplitNoPrimal`. It's not ideal but since backend type parameters are internals, the only thing I can do is wait for translators to be part of your API.

**Comment by wsmoses** [comment]
I think we’re still waiting for feedback from Valentin here.

otherwise I personally say change it to be consistent with WithPrimal then merge it

**Comment by wsmoses** [comment]
Re DI, perhaps in the interim you should throw an error in those cases (rather than risk incorrect results)

**Comment by gdalle** [comment]
I can't easily throw an error in these cases. Indeed, detecting these cases would require accessing the mode's type parameters, which are Enzyme internals (that's why we made the mode setters in the first place).

**Comment by gdalle** [comment]
@wsmoses I renamed `split_mode` to `Split` and `combined_mode` to `Combined` as requested. Do you still want to wait for @vchuravy?

**Review by vchuravy — APPROVED** [review]
Looks reasonable to me.

**Comment by gdalle** [comment]
@wsmoses is this good to go?

**Comment by gdalle** [comment]
Thanks!

**Pull request merged**
